DynamicTablesPkg/Library/Acpi/X64/AcpiMadtLib/MadtGenerator.c

```diff
@@ -0,0 +1,593 @@
+                              CfgMgrProtocol,
+                              CM_NULL_TOKEN,
+                              &IntrSourceOverride,
+                              &IntrSourceOverrideCount
+                              );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: MADT: Failed to get Interrupt Source Override Info object. Status = %r\n",
+      Status
+      ));
+    ASSERT_EFI_ERROR (Status);
+    return Status;
+  }
+
+  if ((IntrSourceOverride == NULL) || (IntrSourceOverrideCount == 0)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: MADT: Interrupt Source Override Info object not found.\n"
+      ));
+    Status = EFI_NOT_FOUND;
+    ASSERT_EFI_ERROR (Status);
+    return Status;
+  }
+
+  for (Index = 0; Index < IntrSourceOverrideCount; Index++) {
+    if ((IntrSourceOverride[Index].Flags &
+         ~(UINT32)(EFI_ACPI_6_5_POLARITY|EFI_ACPI_6_5_TRIGGER_MODE)) != 0)
+    {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: MADT: Invalid Interrupt Source Override Info object. Flags = 0x%x\n",
+        IntrSourceOverride[Index].Flags
+        ));
+      Status = EFI_INVALID_PARAMETER;
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+  }
+
+  LocalApicNmiCount = 0;
+  Status            = GetEX64ObjLocalApicX2ApicNmiInfo (
+                        CfgMgrProtocol,
+                        CM_NULL_TOKEN,
+                        &LocalApicNmi,
+                        &LocalApicNmiCount
+                        );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: MADT: Failed to get Local APIC NMI Info object. Status = %r\n",
+      Status
+      ));
+    ASSERT_EFI_ERROR (Status);
+    return Status;
+  }
+
+  if ((LocalApicNmi == NULL) || (LocalApicNmiCount == 0)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: MADT: Local APIC NMI Info object not found.\n"
+      ));
+    Status = EFI_NOT_FOUND;
+    ASSERT_EFI_ERROR (Status);
+    return Status;
+  }
+
+  for (Index = 0; Index < LocalApicNmiCount; Index++) {
+    if ((LocalApicNmi[Index].Flags & ~(UINT32)(EFI_ACPI_6_5_POLARITY|EFI_ACPI_6_5_TRIGGER_MODE)) != 0) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: MADT: Invalid Local APIC NMI Info object. Flags = 0x%x\n",
+        LocalApicNmi[Index].Flags
+        ));
+      Status = EFI_INVALID_PARAMETER;
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+
+    /// check for AcpiProcessorUid overflow against UINT8_MAX if in APIC mode
+    if (ApicMode == LocalApicModeXApic) {
+      if (LocalApicNmi[Index].AcpiProcessorUid > MAX_UINT8) {
+        DEBUG ((
+          DEBUG_ERROR,
+          "ERROR: MADT: Invalid Local APIC NMI Info object. AcpiProcessorUid = 0x%x\n",
+          LocalApicNmi[Index].AcpiProcessorUid
+          ));
+        Status = EFI_INVALID_PARAMETER;
+        ASSERT_EFI_ERROR (Status);
+        return Status;
+      }
+    }
+  }
+
+  if (ApicMode == LocalApicModeXApic) {

```

Comment:
Maybe @samimujawar has a different view, but I still think it would be better/easier to do the size computation only once, like with:

```
UINT64  MadtSize64;
UINT32  MadtSize;

MadtSize64 = sizeof (EFI_ACPI_6_5_MULTIPLE_APIC_DESCRIPTION_TABLE_HEADER) +
		     (sizeof (EFI_ACPI_6_5_PROCESSOR_LOCAL_APIC_STRUCTURE) * LocalApicCount) +
		     (sizeof (EFI_ACPI_6_5_IO_APIC_STRUCTURE) * IoApicCount) +
		     (sizeof (EFI_ACPI_6_5_INTERRUPT_SOURCE_OVERRIDE_STRUCTURE) * IntrSourceOverrideCount);
if (ApicMode == LOCAL_APIC_MODE_XAPIC) {
  MadtSize64 += (sizeof (EFI_ACPI_6_5_LOCAL_APIC_NMI_STRUCTURE) * LocalApicNmiCount);
} else {
  MadtSize64 += (sizeof (EFI_ACPI_6_5_LOCAL_X2APIC_NMI_STRUCTURE) * LocalApicNmiCount);
}

if (MadtSize64 > MAX_UINT32) {
  DEBUG ((DEBUG_ERROR, "ERROR: MADT: AcpiMadtSize overflow detected.\n"));
  Status = EFI_INVALID_PARAMETER;
  ASSERT_EFI_ERROR (Status);
  return Status;
}

MadtSize = (UINT32)MadtSize64;
```

Reply:
Hi @pierregondois , I have added a macro to make it more readable.